src/Auth/AuthSchemeResolver.php

```diff
@@ -0,0 +1,153 @@
+ */
+class AuthSchemeResolver implements AuthSchemeResolverInterface
+{
+    const UNSIGNED_BODY = '-unsigned-body';
+
+    /**
+     * @var string[] Default mapping of modeled auth trait auth schemes
+     *               to the SDK's supported signature versions.
+     */
+    private static $defaultAuthSchemeMap = [
+        'aws.auth#sigv4' => 'v4',
+        'aws.auth#sigv4a' => 'v4a',
+        'smithy.api#httpBearerAuth' => 'bearer',
+        'smithy.auth#noAuth' => 'anonymous'
+    ];
+
+    /**
+     * @var array Mapping of auth schemes to signature versions used in
+     *            resolving a signature version.
+     */
+    private $authSchemeMap;
+    private $tokenProvider;
+    private $credentialProvider;
+
+
+    public function __construct(
+        callable $credentialProvider,
+        callable $tokenProvider = null,
+        array $authSchemeMap = []
+    ){
+        $this->credentialProvider = $credentialProvider;
+        $this->tokenProvider = $tokenProvider;
+        $this->authSchemeMap = empty($authSchemeMap)
+            ? self::$defaultAuthSchemeMap
+            : $authSchemeMap;
+    }
+
+    /**
+     * Accepts a priority-ordered list of auth schemes and an Identity
+     * and selects the first compatible auth schemes, returning a normalized
+     * signature version.  For example, based on the default auth scheme mapping,
+     * if `aws.auth#sigv4` is selected, `v4` will be returned.
+     *
+     * @param array $authSchemes
+     * @param $identity
+     *
+     * @return string
+     * @throws UnresolvedAuthSchemeException
+     */
+    public function selectAuthScheme(
+        array $authSchemes,
+        array $args = []
+    ): string
+    {
+        $failureReasons = [];
+
+        foreach($authSchemes as $authScheme) {
+            $normalizedAuthScheme = isset($this->authSchemeMap[$authScheme])
+                ? $this->authSchemeMap[$authScheme]
+                : $authScheme;
+
+            if ($this->isCompatibleAuthScheme($normalizedAuthScheme)) {
+                if ($normalizedAuthScheme === 'v4' && !empty($args['unsigned_payload'])) {
+                    return $normalizedAuthScheme . self::UNSIGNED_BODY;
+                }
+
+                return $normalizedAuthScheme;
+            } else {
+                $failureReasons[] = $this->getIncompatibilityMessage($authScheme);
+            }
+        }
+
+        throw new UnresolvedAuthSchemeException(
+            'Could not resolve an authentication scheme: '
+            . implode('; ', $failureReasons)
+        );
+    }
+
+    /**
+     * Determines compatibility based on either Identity or the availability
+     * of the CRT extension.
+     *
+     * @param $authScheme
+     *
+     * @return bool
+     */
+    private function isCompatibleAuthScheme($authScheme): bool
+    {
+        switch ($authScheme) {
+            case 'v4':
+            case 'anonymous':
```

Comment:
Does `anonymous` auth require AWS credentials? Does it require any credentials at all?

Reply:
It technically doesn't, but [`AnonymousSignature`](https://github.com/aws/aws-sdk-php/blob/master/src/Signature/AnonymousSignature.php#L15C5-L20C6) requires an instance of `CredentialsInterface` for signing even though it isn't used.  When customers set `credentials` to `false` in a client constructor, a dummy credentials object is created.

Reply:
Should there be a separate instance then, one specific for anonymous?

Reply:
Ideally.  Currently it doesn't matter if a customer sets credentials to `false` or not— the SDK will accept whatever credentials are available to pass into the `AnonymousSignature` signing functions (e.g. in the case of `authtype` being set to `none`).  To remain backward compatible, we still have to accept `AwsCredentialIdentity` here and an instance of `CredentialsInterface` in `AnonymousSignature`'s signing functions, which makes the implementation of `AnonymousCredentials`, `AnonymousIdentity`, etc. seem like a lot of effort for minimal gain.

Reply:
This case statement is saying that `anonymous` only accepts AWS credentials though, and couldn't be provided bearer credentials (even if ignored). Maybe this should just be a `return true` case on its own?

Reply:
>This case statement is saying that anonymous only accepts AWS credentials though, and couldn't be provided bearer credentials (even if ignored).

As nonsensical as it seems that is actually true.  Bearer takes a different path than other signature types and `anonymous` is in the group of [signatures that require an instance of `CredentialsInterface`](https://github.com/aws/aws-sdk-php/blob/29a79bac02997f3053559f6961a0e83622a14f88/src/Middleware.php#L151,L159).

Reply:
To close the loop, this is done [when `_apply_credentials`](https://github.com/aws/aws-sdk-php/blob/master/src/ClientResolver.php#L586) gets a value of `false` and creates an empty `Credentials` object, which this PR is adjusting to implement the `AwsCredentialIdentity` interface to pass this check.

Reply:
Exactly- that's the explicitly configured case, and there's another where [`anonymous` could be set via `authtype`](https://github.com/aws/aws-sdk-php/blob/master/src/AwsClient.php#L426,L428) and a non-empty `Credentials` object could be used.  

Just for the sake of clarity, (bearer)`Token` and `TokenInterface` are not related to `Credentials` and `CredentialsInterface`, which is why we're checking for `AwsCredentialIdentity`